staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -18,20 +18,37 @@ package shufflesharding
 
 import (
 	"errors"
+	"fmt"
 	"math"
+	"strings"
 )
 
 const maxHashBits = 60
 
-// ValidateParameters can validate parameters for shuffle sharding
-// in a fast but approximate way, including deckSize and handSize
-// Algorithm: maxHashBits >= bits(deckSize^handSize)
-func ValidateParameters(deckSize, handSize int32) bool {
-	if handSize <= 0 || deckSize <= 0 || handSize > deckSize {
-		return false
+// ValidateParameters finds errors in the parameters for shuffle
+// sharding.  Returns an empty slice iff there are no errors.  The
+// entropy requirement is evaluated in a fast but approximate way:
+// bits(deckSize^handSize).
+func ValidateParameters(deckSize, handSize int32) (errs []string) {
+	if handSize <= 0 {
+		errs = append(errs, "handSize is not positive")
 	}
-
-	return math.Log2(float64(deckSize))*float64(handSize) <= maxHashBits
+	if deckSize <= 0 {
+		errs = append(errs, "deckSize is not positive")
+	}
+	if len(errs) > 0 {
+		return
```

Comment:
Could we combine some **if** to make return in **if**?

Reply:
We could, but I chose to use the existing structure because it made a certain amount of sense.  First, we check the range of each individual value, accumulating errors.  If there are any then take an early out.  Then check logically necessary relationships.  Early out if that revealed problems.  Finally do the more complicated and costly quantitative check.

Reply:
i agree with Mike, the current flow is easier to understand.

Reply:
@yliaog : I am not sure which you are agreeing with.  Is the revised code OK?

Reply:
OK，let's keep this.